{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'timestamp': datetime.datetime(2025, 6, 10, 22, 5, 16, 730982), 'city': 'Екатеринбург', 'source': 'AccuWeather', 'T0': 28, 'P0': 760.0, 'H0': 41, 'Ff': 7, 'WD0': 'З', 'UVI': 0, 'PrecipitationType': 'Преимущественно облачно', 'AQI': 51, 'PM2.5': 51, 'PM10': 22, 'NO2': 51, 'O3': 32, 'CO': 2, 'SO2': 1}\n",
      "{'timestamp': datetime.datetime(2025, 6, 10, 22, 5, 18, 156050), 'city': 'Екатеринбург', 'source': 'AccuWeather', 'TM1': None, 'TD1': 31, 'TE1': None, 'TN1': 18, 'PM1': None, 'PD1': None, 'PE1': None, 'PN1': None, 'HM1': None, 'HD1': None, 'HE1': None, 'HN1': None, 'WSM1': None, 'WSD1': 13, 'WSE1': None, 'WSN1': 9, 'WDM1': None, 'WDD1': 'Ю', 'WDE1': None, 'WDN1': 'ЮЮВ', 'PrM1': None, 'PrD1': 'Солнечно', 'PrE1': None, 'PrN1': 'Преимущественно ясно', 'MaxUVI1': 7, 'MinUVI1': None}\n"
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the current cell or a previous cell. \n",
      "\u001b[1;31mPlease review the code in the cell(s) to identify a possible cause of the failure. \n",
      "\u001b[1;31mClick <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. \n",
      "\u001b[1;31mView Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "import requests, random\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import datetime as dt\n",
    "import time, re\n",
    "from bs4 import BeautifulSoup\n",
    "import datetime as dt\n",
    "from providers.weather_provider import WeatherProvider\n",
    "\n",
    "        \n",
    "class TestProvider(WeatherProvider):\n",
    "    def __init__(self):\n",
    "        super().__init__()\n",
    "        self.provider_name = \"AccuWeather\"\n",
    "        self.base_url = \"https://www.accuweather.com\"\n",
    "\n",
    "    def get_location_key(self, city):\n",
    "        url = f\"{self.base_url}/web-api/autocomplete\"\n",
    "        params = {\"query\": city, \"language\": \"ru\"}\n",
    "        resp = self.session.get(url, params=params)\n",
    "        data = resp.json()\n",
    "        if not data:\n",
    "            raise ValueError(f\"[AccuWeather] Город {city} не найден\")\n",
    "        return data[0]['key']\n",
    "\n",
    "    def get_city_url(self, loc_key:str):\n",
    "        url = f\"{self.base_url}/web-api/three-day-redirect?key={loc_key}\"\n",
    "        current_resp = self.session.get(url, timeout=10)\n",
    "        current_soup = BeautifulSoup(current_resp.text, \"html.parser\")\n",
    "        return self.base_url + current_soup.select_one(\"a.cur-con-weather-card\")['href']\n",
    "        \n",
    "    def fetch(self, city: str) -> dict:\n",
    "        location_key = self.get_location_key(city)\n",
    "\n",
    "        current_url = self.get_city_url(location_key)\n",
    "        current_resp = self.session.get(current_url, timeout=10)\n",
    "        current_soup = BeautifulSoup(current_resp.text, \"html.parser\")\n",
    "        \n",
    "        # Текущая температура\n",
    "        try:\n",
    "            temp_elem = current_soup.select_one(\"div.temp div.display-temp\")\n",
    "            temp_str = temp_elem.get_text(strip=True)\n",
    "            temp_now = self._safe_int(temp_str) if temp_str else None\n",
    "        except Exception as e:\n",
    "            temp_now = None\n",
    "            print(f\"[{self.provider_name}] Ошибка чтения текущей температуры. Исключение: {e}\")\n",
    "            \n",
    "        # Текущий тип погодных условий\n",
    "        try:\n",
    "            precip_elem = current_soup.select_one('div.current-weather div.phrase')\n",
    "            precips_now = precip_elem.get_text(strip=True) if precip_elem else None\n",
    "        except Exception as e:\n",
    "            precips_now = None\n",
    "            print(f\"[{self.provider_name}] Ошибка чтения текущего типа погодных условий. Исключение: {e}\")\n",
    "        \n",
    "        uv_now = 0\n",
    "        humidity_now = None\n",
    "        pressure_now = None\n",
    "        wind_speed = None\n",
    "        wind_direction = None\n",
    "        try:    \n",
    "            details_card =  current_soup.find('div', class_='current-weather-card')\n",
    "            items = details_card.find_all('div', class_='detail-item')\n",
    "            for item in items:\n",
    "                if item:\n",
    "                    if \"Влажность\" in item.text:\n",
    "                        humid_str = re.search(r'\\d+', item.text).group()\n",
    "                        humidity_now = self._safe_int(humid_str)\n",
    "                    if \"Давление\" in item.text:\n",
    "                        pressure_str = re.search(r'\\d{3,4}', item.text).group()\n",
    "                        pressure_now = self._safe_int(pressure_str) // 1.333 if pressure_str else None\n",
    "                    if \"Макс. УФ-индекс\" in item.text:\n",
    "                        uv_str = re.search(r'\\d+', item.text).group()\n",
    "                        uv_now = self._safe_int(uv_str) if uv_str else 0\n",
    "                    if \"Ветер\" in item.text:\n",
    "                        wind_speed_str = re.search(r'(\\d+) км/ч', item.text).group(1)\n",
    "                        wind_speed = self._safe_int(wind_speed_str) if wind_speed_str else 0\n",
    "                        wind_direction = re.search(r'([ЗВСЮ]{1,3}) \\d+ км/ч', item.text).group(1)\n",
    "        except Exception as e:\n",
    "            print(f\"[{self.provider_name}] Ошибка чтения текущих влажности, давления и УФ-индекса. Исключение: {e}\")\n",
    "        \n",
    "        # Качество воздуха и загрязнители\n",
    "        quality_url = current_url.replace(\"current-weather\",\"air-quality-index\")\n",
    "        quality_resp = self.session.get(quality_url, timeout=10)\n",
    "        quality_soup = BeautifulSoup(quality_resp.text, 'html.parser')\n",
    "        \n",
    "        aq_number = None\n",
    "        pm25 = None\n",
    "        pm10 = None\n",
    "        no2_gas = None\n",
    "        o3_gas = None\n",
    "        co_gas = None\n",
    "        so2_gas = None\n",
    "        try:\n",
    "            # Индекс качества воздуха\n",
    "            aq_number_elem = quality_soup.select_one('div.aq-number')\n",
    "            aq_number = self._safe_int(aq_number_elem.get_text(strip=True)) if aq_number_elem.get_text(strip=True).isdigit() else None\n",
    "            \n",
    "            # Загрязнители\n",
    "            pollutant_elems = quality_soup.find_all('div', class_='pollutant-index')\n",
    "            pm25 = self._safe_int(pollutant_elems[0].get_text(strip=True))\n",
    "            no2_gas = self._safe_int(pollutant_elems[2].get_text(strip=True))\n",
    "            o3_gas = self._safe_int(pollutant_elems[4].get_text(strip=True))\n",
    "            pm10 = self._safe_int(pollutant_elems[6].get_text(strip=True))\n",
    "            co_gas = self._safe_int(pollutant_elems[8].get_text(strip=True))\n",
    "            so2_gas = self._safe_int(pollutant_elems[10].get_text(strip=True))  \n",
    "        except Exception as e:\n",
    "            print(f\"[{self.provider_name}] Ошибка чтения качества и состава воздуха. Исключение: {e}\")\n",
    "        \n",
    "        # Составляем агрегированные данные\n",
    "        return self.make_dummy(self.provider_name, \n",
    "                               city=city,\n",
    "                               timestamp=None,\n",
    "                               \n",
    "                               temp=temp_now, \n",
    "                               pres=pressure_now, \n",
    "                               hum=humidity_now, \n",
    "                               wind_speed=wind_speed,\n",
    "                               wind_direction=wind_direction,\n",
    "                               uv_index=uv_now,\n",
    "                               precipitationTypes= precips_now,\n",
    "                               air_quality_index= aq_number,\n",
    "                               pm25=pm25, pm10=pm10, no2_gas=no2_gas, o3_gas=o3_gas, co_gas=co_gas, so2_gas=so2_gas,\n",
    "                               )\n",
    "        \n",
    "    def fetch_forecast(self, city: str) -> dict:\n",
    "        location_key = self.get_location_key(city)\n",
    "        \n",
    "        # Данные прогноза\n",
    "        default_url = self.get_city_url(location_key) \n",
    "        forecast_url = default_url.replace(\"current-weather\",\"weather-tomorrow\")\n",
    "        forecast_resp = self.session.get(forecast_url, timeout=10)\n",
    "        forecast_soup = BeautifulSoup(forecast_resp.text, \"html.parser\")\n",
    "        \n",
    "        # Температура (дневная и ночная)\n",
    "        try:\n",
    "            temps_elems = forecast_soup.select(\"div.half-day-card-header__content div.weather div.temperature\")\n",
    "            temps_str = [re.search(r'\\d+',elem.get_text(strip=True)).group() for elem in temps_elems]\n",
    "            temps = [self._safe_int(temp_str) if temp_str and temp_str.isdigit() else None for temp_str in temps_str]\n",
    "            max_temp, min_temp = temps\n",
    "        except Exception as e:\n",
    "            max_temp = min_temp = None\n",
    "            print(f\"[{self.provider_name}] Ошибка чтения максимальной и минимальной температуры воздуха на завтра. Исключение: {e}\")\n",
    "        \n",
    "        # УФ-индекс\n",
    "        max_uv_index = None\n",
    "        try:\n",
    "            elements = forecast_soup.find_all('p', class_='panel-item')\n",
    "            if elements:\n",
    "                for elem in elements:\n",
    "                    if 'Макс. УФ-индекс' in elem.text:\n",
    "                        uv_text = elem.find('span', class_='value').get_text()\n",
    "                        uv_number = re.search(r'\\d+', uv_text).group()\n",
    "                        max_uv_index = self._safe_int(uv_number)\n",
    "                        break\n",
    "        except Exception as e:\n",
    "            print(f\"[{self.provider_name}] Ошибка чтения УФ-индекса на завтра. Исключение: {e}\")\n",
    "            \n",
    "        # Скорость и направление ветра\n",
    "        wind_speed_day = wind_speed_night = None\n",
    "        wind_dir_day = wind_dir_night = None\n",
    "        try:\n",
    "            elements = forecast_soup.find_all('p', class_='panel-item')\n",
    "            wind_speeds = []\n",
    "            wind_dirs = []\n",
    "            if elements:\n",
    "                for elem in elements:\n",
    "                    if 'Ветер' in elem.text:\n",
    "                        wind_text = elem.find('span', class_='value').get_text()\n",
    "                        wind_speed_number = int(re.search(r'\\d+', wind_text).group())\n",
    "                        wind_direction_str = re.search(r'[ВЗСЮ]{1,3}', wind_text).group()\n",
    "                        wind_speeds.append(wind_speed_number)\n",
    "                        wind_dirs.append(wind_direction_str)\n",
    "            wind_speed_day = wind_speeds[0]\n",
    "            wind_speed_night = wind_speeds[1]\n",
    "            wind_dir_day = wind_dirs[0]\n",
    "            wind_dir_night = wind_dirs[1]\n",
    "        except Exception as e:\n",
    "            print(f\"[{self.provider_name}] Ошибка чтения скорости (км/ч) и направления ветра на завтра. Исключение: {e}\")\n",
    "\n",
    "        try:         \n",
    "            precip_elems = forecast_soup.select(\"div.half-day-card-content div.phrase\")\n",
    "            day_precip = precip_elems[0].get_text(strip=True)\n",
    "            night_precip = precip_elems[1].get_text(strip=True)\n",
    "        except Exception as e:\n",
    "            day_precip, night_precip = None\n",
    "            print(f\"[{self.provider_name}] Ошибка чтения типа погодных условий на завтра. Исключение: {e}\")\n",
    "            \n",
    "        \n",
    "        # Составляем агрегированные данные\n",
    "        return self.make_forecast_dummy(self.provider_name, \n",
    "                               city=city,\n",
    "                               timestamp=None,\n",
    "                               \n",
    "                               temp_day=max_temp,\n",
    "                               temp_night=min_temp,\n",
    "                               \n",
    "                               wind_speed_day=wind_speed_day,\n",
    "                               wind_speed_night=wind_speed_night,\n",
    "                               wind_dir_day=wind_dir_day,\n",
    "                               wind_dir_night=wind_dir_night,\n",
    "                               \n",
    "                               precips_day=day_precip,\n",
    "                               precips_night=night_precip,\n",
    "                               \n",
    "                               max_uv_index=max_uv_index\n",
    "                               )\n",
    "        \n",
    "print(TestProvider().fetch(\"Екатеринбург\"))\n",
    "print(TestProvider().fetch_forecast(\"Екатеринбург\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
